src/Aspire.Hosting/Node/NodeExtensions.cs

```diff
@@ -0,0 +1,70 @@
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="workingDirectory">The working directory to use for the command. If null, the working directory of the current process is used.</param>
+    /// <param name="args">The arguments to pass to the command.</param>
+    /// <returns></returns>
+    public static IResourceBuilder<NodeAppResource> AddNodeApp(this IDistributedApplicationBuilder builder, string name, string workingDirectory, string[] args)
+    {
+        var resource = new NodeAppResource(name, "node", workingDirectory, args);
+
+        return builder.AddResource(resource)
+                      .WithNodeDefaults();
+    }
+
+    /// <summary>
+    /// Adds a node application to the application model. Executes the npm command with the specified script name.
+    /// </summary>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/> to add the resource to.</param>
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="workingDirectory">The working directory to use for the command. If null, the working directory of the current process is used.</param>
+    /// <param name="scriptName">The npm script to execute. Defaults to "start".</param>
+    /// <param name="args">The arguments to pass to the command.</param>
+    /// <returns></returns>
+    public static IResourceBuilder<NodeAppResource> AddNpmApp(this IDistributedApplicationBuilder builder, string name, string workingDirectory, string scriptName = "start", string[]? args = null)
+    {
+        string[] allArgs = args is { Length: > 0 }
+            ? ["run", scriptName, "--", .. args]
+            : ["run", scriptName];
+
+        var resource = new NodeAppResource(name, "npm", workingDirectory, allArgs);
+
+        return builder.AddResource(resource)
+                      .WithNodeDefaults();
+    }
+
+    /// <summary>
+    /// Injects service discovery information as environment variables from the project resource into the destination resource, using the source resource's name as the service name.
+    /// Each service binding defined on the project resource will be injected using the format "services__{sourceResourceName}__{bindingIndex}={bindingNameQualifiedUriString}."
+    /// </summary>
+    /// <typeparam name="TDestination">The destination resource.</typeparam>
+    /// <typeparam name="TSource">The source project resource.</typeparam>
+    /// <param name="builder">The resource where the service discovery information will be injected.</param>
+    /// <param name="source">The resource from which to extract service bindings.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{TDestination}"/>.</returns>
+    public static IResourceBuilder<TDestination> WithReference<TDestination, TSource>(this IResourceBuilder<TDestination> builder, IResourceBuilder<TSource> source)
```

Comment:
@mitchdenny Now might be the time to fix this. Right now, this requires adding explicit overloads for things that support service discovery. We need to make it more contract based.